### PR TITLE
[INTEL MKL] Add MklSlice Op

### PR DESCRIPTION
**Pull request by pandaoxin**
Add MklSlice op, please review.

### CODE REVIEW

File: tensorflow/core/kernels/mkl_slice_op.cc [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+      CHECK_EQ(input_mkl_shape.IsMklTensor(), output_mkl_shape.IsMklTensor());
+      output.SetUsrMem(output_md, output_tensor);
+
+      std::vector<primitive> net;
+      // Step 4 - create reorder primitive desc between view_pd and output_pd.
+      auto reorder_pd =
+          reorder::primitive_desc(view_pd, output.GetUsrMemPrimDesc());
+      // Step 5 - create reorder primitive itself.
+      net.push_back(reorder(reorder_pd, *src.GetUsrMem(), *output.GetUsrMem()));
+      stream(stream::kind::eager).submit(net).wait();
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_slice_op.cc]: Please add a comment that this line executes the primitive.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for your PR! 

There is already an MKLSliceOp which replaces TF's Slice by default. How is MKLDnnSliceOp going to be used? Is it meant to replace MKLSliceOp eventually, or are you just going to use it in some special cases? 

Please also add some tests for the op.

File: tensorflow/core/kernels/mkl_slice_op.cc [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+    int64 b = (*begin)[i];
+    int64 s = (*size)[i];
+    if (input_tf_shape.dim_size(i) == 0) {
+      OP_REQUIRES(
+          context, b == 0 && s == 0,
+          errors::InvalidArgument("Expected begin[", i, "] == 0 (got ", b,
+                                  ") and size[", i, "] == 0 ", "(got ", s,
+                                  ") when ", "input.dim_size(", i, ") == 0"));
+    } else {
+      OP_REQUIRES(context, 0 <= b && b <= input_tf_shape.dim_size(i),
```
[the remark this answers is not in the document]
Comment (pandaoxin) [reply tensorflow/core/kernels/mkl_slice_op.cc]: Ok.

**Comment by pandaoxin** [comment]
@penpornk Thanks for your detailed comments! Have modified the code. This patch uses Mkl-dnn's primitive to implement slice, while the Mklslice already exists works for Intel Mkl.

File: tensorflow/core/ops/array_ops.cc [hunk 3]
```diff
@@ -1681,8 +1681,8 @@ REGISTER_OP("_MklSlice")
       TF_RETURN_IF_ERROR(c->MakeShapeFromShapeTensor(1, &begin_value));
 
       // NOTE(mrry): We can't use `MakeShapeFromShapeTensor` for `sizes` because
-      // it might contain -1, which can't be represented -1 in the ShapeHandle
-      // would meqan "unknown".
+      // it might contain -1, which can't be represented. (-1 in the ShapeHandle
+      // would mean "unknown".)
       const Tensor* sizes_value = c->input_tensor(3);
```
Comment (penpornk) [inline tensorflow/core/ops/array_ops.cc]: If this is the only difference from the lambda shape function in "Slice", please refactor the function and put it in [common_shape_fns.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/common_shape_fns.cc) instead. Please see [_MklConcat](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/array_ops.cc#L3059) for example.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the changes and the clarification! I think we should refactor the lambda shape function out to common_shape_fns.cc since it's almost the same as the original one in TF Slice op. Please also add tests for this op.

**Comment by pandaoxin** [comment]
Thank you @penpornk ! I have modified the location of Slice shape function. And I also remember that there is a test added in mkl_layout_pass test(tensorflow/core/graph/mkl_layout_pass_test.cc). Besides, there already exists some UT tests for Slice op.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you @pandaoxin! I have some last minor changes. Please also run clang-format. It seems there are some extra blank lines. 

Tests: My bad. Sorry about that! For some reason I forgot about the test in `mkl_layout_pass_test.cc`.

**Comment by pandaoxin** [comment]
Hi @penpornk , thanks for your advice! Modifications done.

**Review by penpornk — DISMISSED** [review]
Thank you again for the PR!

**Comment by martinwicke** [comment]
@pandaoxin can you look into the lint errors?

**Comment by pandaoxin** [comment]
@martinwicke The failed test said that:
'Missing dependency: //tensorflow/contrib/eager/python:parameter_server',
which seems not related to my committed code, while the lint error as well(cannot find my file).
Could you help me check the error? Thanks very much!

**Comment by martinwicke** [comment]
Yeah, that does look unrelated. I'll try again, maybe it was broken at head.

**Comment by martinwicke** [comment]
@pandaoxin Can you change the CHECKs for either DCHECKs or even better proper errors? We're trying to remove CHECKs since they have the ability to kill production serving system, while errors (non-OK Status) can be handled gracefully.

**Comment by pandaoxin** [comment]
@martinwicke Ok, thank you! Code modified.

**Comment by penpornk** [comment]
Thank you! Please also change `CHECK_NOTNULL` to `DCHECK_NOTNULL` and `LOG(FATAL)` to `LOG(DFATAL)` (in `mkl_slice_op.cc`).

**Comment by pandaoxin** [comment]
Hi @penpornk , since there is neither DCHECK_NOTNULL nor DFATAL, I turn both of them into DCHECK. Please review.

**Comment by penpornk** [comment]
There might be a better substitute. Let's wait for @martinwicke.

**Comment by martinwicke** [comment]
DCHECK for both these is fine.

**Review by penpornk — APPROVED** [review]

**Comment by pandaoxin** [comment]
Thank you! @penpornk @martinwicke

**Comment by pandaoxin** [comment]
@penpornk @martinwicke Merge button still unavailable, seems two approving reviews required?

**Comment by martinwicke** [comment]
It's making its way through the infra. The merge button is meaningless -- we use our scripts to merge anyway.

**Comment by pandaoxin** [comment]
@martinwicke ok, thanks.

**Pull request merged**